### PR TITLE
feat: Reflow and Resize while Dragging and Resizing widgets.

### LINKED ISSUE #2308: [Epic] Reflow widgets
## Problem
The problems we want to solve in this feature are
1. Users dragging/resizing widgets find it difficult to place the widget at a location that does not fit the widget size or is currently occupied by another widget.
2. Users who are trying to build UI that contains repetitive elements find it difficult because the copy-paste functionality pastes the widget at an unexpected/unintelligent location. It is also difficult to correct the pasted widget location and there is no way to specify the exact widget location that a user wants to place the widget.

## Requirements

| Title | Desc | Priority | Comment |
| ----- | ------ | ----- | ------ | 
Drag/Drop | It should be easy to drag / drop a widget anywhere on the canvas even if there is little to no space | High | - 
Drag/Resize | It should be easy to resize a widget to the size a user envisions without having to worry about the surrounding widgets | High | -

## Out of scope:
- Dynamic height of widgets
- Reflowing mobile view
- Copy Paste #5150

### Drag & Drop
To enable eas of use for drag & drop, we explored the following ideas
- **Enable Layers(Hard):** Stack multiple widgets on top of each other. Leave it to the user to handle the correct placement and stacking of widgets. We don't prefer this option because
  - While it is predictable, It does not help the user arrange the widgets and users must painstakingly group/move widgets around to achieve the optimal position
  - It requires a larger canvas to be able to move stacked widgets out of the way and would be a better experience with an infinite canvas
  - It is a new construct and has certain implementation challenges
- **Reflow:** When a widget (dragged) collides with another widget on the canvas (collided), we can move the surrounding widgets to accommodate the new widget. We don't prefer this option because it is hard to predict the outcome of the drag and more importantly difficult to control because the only way to go back to the previous state is to drop the widget back to its original dragged location
- **Resize (Preferred):** When a widget collides with another widget on the canvas, we can resize the widget to fit in the available space on the canvas thus ensuring the drop will always be successful. While this may not be the ideal size of the widget, we prefer this approach because it's predictable as the user can see the drop target resize and controllable since the user can resize the widget to the ideal size after.

### Drag & Resize
While resizing we think it would be a good experience to move the widgets colliding with the resized widget out of the way. To achieve this, there are a few options we explored
-  **Vertical Reflow(Easy):** When a widget collides with another widget on the canvas, the widget on the canvas and all widgets below it are pushed below the resized widget. The relative distances between widgets being pushed must remain the same. 
- **Push Reflow:** When a widget collides with another widget, it continues to push it in the direction of its push(resize). If the widget encounters another widget, this widget also begins to move in the same direction till the last widget encounters a boundary. Pushing a widget into a boundary causes it to move below the widget pushing it to give it space and moves all widgets below as well using the same interaction.
- **Relative Push Reflow(Preferred):** This works similar to the above interaction except when a widget is pushed, it immediately pushed all widgets in the same direction maintaining their relative spaces (white-space) till they encounter a boundary. After they encounter a boundary, the widgets collapse the spaces in order of the spaces closest to farthest from the boundary. Once no more spaces are present, the widget closest to the boundary moves below the widget pushing it to give it space, and moves all widgets below as well maintaining the same relative spacing. This is the preferred method to ensure spaces intentionally created by a user below the current fold of the screen are not distorted without the user being aware of them.

### Callouts
- All interactions will only be valid within a container
- Movement of surrounding widgets will be dynamic as the resize occurs but will NOT be permanent till the user ends the resize interaction
- All interactions should be undo/redo able

### Copy Paste
The default mode of copy-paste will be to place a widget in the first available space and reflow the surrounding widgets if no space is available using the collision algorithm above. The first available space can be defined by the first space (Right -> Below -> Left ->Top) of the current widget in the order specified. Additional rules on top of these can be added for specific cases like alignment.
Users must have an option to copy + select where to paste the widget (Design to help with this)

## Deploy Preview
relative
https://deploy-preview-9055--appsmith-editor-ce.netlify.app/
natural
https://deploy-preview-9054--appsmith-editor-ce.netlify.app/

## RACI Matrix 

| Attribute | Assignee | 
| ----- | ----- |
| Responsible | @marks0351, @rahulramesha | 
| Accountable | @marks0351 | 
| Consulted | @Nikhil-Nandagopal, @riodeuno  @areyabhishek | 
| Informed | @somangshu |

### LINKED ISSUE #7262: [Feature] Resizing widget downward is very janky
## Summary

This majorly is an issue when the widget is placed at the bottom of the canvas, The resize handle does not autoscroll to increase the size of the widget. The interaction need to be smooth here

### LINKED ISSUE #7719: [Feature] Drag and Resize Reflow - POC

### LINKED ISSUE #7720: [Feature] Resize and Reflow
Feature for Resize and reflow Algorithm.

### LINKED ISSUE #7850: [Feature] Drag and Resize Reflow - Cypress & Jest Tests

### PR DESCRIPTION
**Pull request by marks0351**
### Terminology: 

Target widget - widget that is moving / widget that is resizing
Reflowing widget - widget that is moved by the Reflow algorithm because it was colliding with the target widget.
Droppable Widget - Widgets that can contain other widgets (Container widget, Tabs Widget, List Widget)
Container Jump - Target widget entering/exiting a Droppable widget

### What's Reflow: 

Appsmith’s Reflow is about moving widgets out of the way and providing space to drop/resize a widget.

When the target widget comes into contact with any other widget it moves(reflows) until it hits the boundary of the container and resizes the widget once it hits the boundary of its container.

While dragging the target widget, direction of mouse movement is taken into account to reflow widgets i.e, if the target widget hits another widget from the left -> right direction it moves the reflowing widget in the same direction. So mouse movement direction can be used to reflow widgets in different directions.

#### Note:
*If the Reflowing widget is a Droppable widget, then mouse speed is taken into account to differentiate Container jumps and reflowing the Droppable widget.
Move the target widget faster to do a container jump, move the target widget slowly to reflow the widget.

### Reflow Algorithm(what happens under the hood):

Let us consider the widget being dragged on canvas as a Static widget. 
When a Static widget is dragged on canvas, we compare its coordinates with the coordinates of all other widgets on the canvas to check for overlapping collisions.
The Overlapping widgets are further put through the same process recursively to get the widgets overlapping with those widgets.
This helps in creation of a Tree structure of widgets, wherein a parent node will have overlapping widgets as children nodes and those in turn become parents for their overlapping widgets.
With the help of this Tree structure, the direction of the static widget, displacement of the static widget and canvas boundaries, X and Y movement values of each widget is calculated.
These widgets when moved along the X and Y axis from their original position will create the illusion of pushing the colliding widgets.
This is the core logic on how the algorithm works, but it has a lot more logic under the hood, like tweaking direction of movement in corner cases, keeping track of multiple directions of widgets, smooth canvas exits and entries etc.

### Implementation: 
We have added a redux state to capture reflowed widgets when any single widget is being dragged or resized.
position of reflowed widgets is updated based on the reflowed widgets state.
react hooks are used to get triggers of mouse moves and update reflowed widgets state based on the algorithm mentioned above.

Fixes #2308
Fixes #7262
Fixes #7719 
Fixes #7720
Fixes #7850

## Type of change

> Please delete options that are not relevant.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

## How Has This Been Tested?

> Please describe the tests that you ran to verify your changes. Provide instructions, so we can reproduce.
> Please also list any relevant details for your test configuration.

- Test A
- Test B

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:green_circle: Total coverage has increased</summary>

    // Code coverage diff between base branch:release and head branch: reflow-natural-with-drag 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 55.01 **(0)** | 36.72 **(0.23)** | 34.7 **(0.22)** | 55.51 **(0.01)**
 :sparkles: :new: | **app/client/src/actions/reflowActions.ts** | **70** | **100** | **0** | **57.14**
 :red_circle: | app/client/src/components/designSystems/appsmith/PositionedContainer.tsx | 100 **(0)** | 77.42 **(-10.08)** | 100 **(0)** | 100 **(0)**
 :green_circle: | app/client/src/components/editorComponents/DropTargetComponent.tsx | 89.19 **(0.78)** | 69.23 **(3.27)** | 90.91 **(0.91)** | 88.89 **(0.83)**
 :green_circle: | app/client/src/components/editorComponents/ResizableComponent.tsx | 64.37 **(6.04)** | 19.67 **(-2.17)** | 58.33 **(-8.34)** | 64.37 **(5.49)**
 :red_circle: | app/client/src/pages/Editor/MainContainerLayoutControl.tsx | 83.33 **(-4.17)** | 57.58 **(-17.42)** | 57.14 **(-9.53)** | 83.33 **(-4.17)**
 :green_circle: | app/client/src/pages/common/CanvasDraggingArena.tsx | 100 **(0)** | 90 **(2.5)** | 100 **(0)** | 100 **(0)**
 :green_circle: | app/client/src/pages/common/CanvasSelectionArena.tsx | 91.58 **(0)** | 72.13 **(0.46)** | 89.29 **(0)** | 91.3 **(0)**
 :sparkles: :new: | **app/client/src/reducers/uiReducers/reflowReducer.ts** | **44.44** | **100** | **0** | **44.44**
 :sparkles: :new: | **app/client/src/reflow/index.ts** | **65** | **66.67** | **100** | **68.75**
 :sparkles: :new: | **app/client/src/reflow/reflowHelpers.ts** | **3.76** | **0** | **0** | **4.5**
 :sparkles: :new: | **app/client/src/reflow/reflowTypes.ts** | **100** | **100** | **100** | **100**
 :sparkles: :new: | **app/client/src/reflow/reflowUtils.ts** | **75** | **63.64** | **79.31** | **76.47**
 :sparkles: :new: | **app/client/src/resizable/resize/index.tsx** | **10.71** | **4.35** | **0** | **12.33**
 :sparkles: :new: | **app/client/src/resizable/resizenreflow/index.tsx** | **54.01** | **49.53** | **40** | **52.34**
 :red_circle: | app/client/src/sagas/DraggingCanvasSagas.ts | 76 **(-22.57)** | 53.85 **(-17.58)** | 72.73 **(-27.27)** | 77.78 **(-20.68)**
 :sparkles: :new: | **app/client/src/sagas/ReflowSagas.ts** | **36.36** | **11.11** | **50** | **40**
 :green_circle: | app/client/src/sagas/WidgetAdditionSagas.ts | 69.57 **(0.23)** | 60 **(0)** | 62.5 **(0)** | 70.49 **(0.24)**
 :red_circle: | app/client/src/sagas/WidgetOperationSagas.tsx | 51.32 **(-2.67)** | 38.54 **(-4.16)** | 52.08 **(-1.11)** | 52.79 **(-2.8)**
 :green_circle: | app/client/src/sagas/index.tsx | 92.86 **(0.13)** | 60 **(0)** | 100 **(0)** | 94.44 **(0.1)**
 :green_circle: | app/client/src/selectors/editorSelectors.tsx | 76.29 **(1.57)** | 70.13 **(3.46)** | 61.54 **(3.91)** | 74.05 **(2.52)**
 :sparkles: :new: | **app/client/src/selectors/widgetReflowSelectors.tsx** | **100** | **80** | **100** | **100**
 :red_circle: | app/client/src/utils/DSLMigrations.ts | 73.4 **(-0.71)** | 64.27 **(-0.67)** | 70.21 **(0)** | 73.21 **(-0.73)**
 :red_circle: | app/client/src/utils/WidgetPropsUtils.tsx | 86.42 **(-0.64)** | 76 **(0)** | 83.33 **(-1.29)** | 83.08 **(-0.74)**
 :sparkles: :new: | **app/client/src/utils/reflowHookUtils.ts** | **68.75** | **66.67** | **62.5** | **67.74**
 :red_circle: | app/client/src/utils/storage.ts | 34.09 **(0.13)** | 13.04 **(-2.75)** | 7.89 **(-1.49)** | 26.92 **(0.87)**
 :red_circle: | app/client/src/utils/autocomplete/TernServer.ts | 52.71 **(-0.23)** | 40.83 **(-0.84)** | 36.21 **(0)** | 56.74 **(-0.25)**
 :red_circle: | app/client/src/utils/hooks/useBlocksToBeDraggedOnCanvas.ts | 92.48 **(-3.75)** | 87.21 **(-4.98)** | 82.76 **(-2.95)** | 93.7 **(-2.38)**
 :red_circle: | app/client/src/utils/hooks/useCanvasDragging.ts | 86.79 **(-9.76)** | 68.29 **(-9.95)** | 87.88 **(-2.6)** | 86.73 **(-9.75)**
 :sparkles: :new: | **app/client/src/utils/hooks/useReflow.ts** | **80** | **57.14** | **100** | **80.85**
 :x: | ~~app/client/src/resizable/index.tsx~~ | ~~59.52~~ | ~~52.17~~ | ~~35~~ | ~~61.64~~</details>

### CODE REVIEW

File: app/client/src/actions/reflowActions.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
+import {
+  ReduxAction,
+  ReflowReduxActionTypes,
+} from "constants/ReduxActionConstants";
+import { widgetReflowOnBoardingState } from "reducers/uiReducers/reflowReducer";
+import { ReflowedSpaceMap } from "reflow/reflowTypes";
+
+export const reflowMove = (
```
Comment (marks0351) [inline app/client/src/actions/reflowActions.ts:8]: @rahulramesha lets suffix these with the work action

File: app/client/src/pages/common/CanvasDraggingArena.tsx [hunk 2]
```diff
@@ -73,3 +76,6 @@ export function CanvasDraggingArena({
   ) : null;
 }
 CanvasDraggingArena.displayName = "CanvasDraggingArena";
+// CanvasDraggingArena.whyDidYouRender = {
```
Comment (marks0351) [inline app/client/src/pages/common/CanvasDraggingArena.tsx:79]: @marks0351 need to remove these comments

File: app/client/src/reflow/index.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+  immediateExitContainer?: string,
+  prevPositions?: OccupiedSpace,
+  prevCollidingSpaces?: CollidingSpaceMap,
+) {
+  const isHorizontalMove = getIsHorizontalMove(newPositions, prevPositions);
+  const filteredOccupiedSpace = filterSpaceById(
+    newPositions.id,
+    occupiedSpaces,
+  );
+  const consolelog = cloneDeep({
```
Comment (marks0351) [inline app/client/src/reflow/index.ts:31]: @rahulramesha do we need this anymore?

File: app/client/src/reflow/index.ts [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  if (!isColliding || !OGPositions || direction === ReflowDirection.UNSET) {
+    return {
+      movementLimit: {
+        canHorizontalMove: true,
+        canVerticalMove: true,
+      },
+    };
+  }
+  //eslint-disable-next-line
+  console.log("reflow input", consolelog);
```
Comment (marks0351) [inline app/client/src/reflow/index.ts:63]: @rahulramesha lets remove this

File: app/client/src/reflow/reflowHelpers.ts [hunk 5]
```diff
@@ -0,0 +1,450 @@
...
+    occupiedSpaces,
+    newPositions,
+    collidingSpaceMap,
+  );
+
+  if (
+    !collisionTree ||
+    !collisionTree.children ||
+    Object.keys(collisionTree.children).length <= 0
+  )
```
Comment (marks0351) [inline app/client/src/reflow/reflowHelpers.ts:48]: @rahulramesha can we please add {} for most if else blocks to provide readability and consistency.

File: app/client/src/reflow/reflowHelpers.ts [hunk 6]
```diff
@@ -0,0 +1,450 @@
...
+      isHorizontal: accessors.isHorizontal,
+    });
+  }
+  const newPositionsMovement = {
+    id: collisionTree.id,
+    directionalMovements,
+  };
+
+  //eslint-disable-next-line
+  console.log("reflowInfo", {
```
Comment (marks0351) [inline app/client/src/reflow/reflowHelpers.ts:133]: lets remove the console.log

File: app/client/src/resizable/resizenreflow/index.tsx [hunk 7]
```diff
@@ -0,0 +1,507 @@
...
+  enable: boolean;
+  className?: string;
+  parentId?: string;
+  widgetId: string;
+  gridProps: GridProps;
+  zWidgetType?: string;
+  zWidgetId?: string;
+};
+
+export function Resizable(props: ResizableProps) {
```
Comment (marks0351) [inline app/client/src/resizable/resizenreflow/index.tsx:158]: Can we prefix it with Reflow to differentiate the generic resizable and reflow resizable.

File: app/client/src/sagas/WidgetOperationSagas.tsx [hunk 8]
```diff
@@ -147,6 +163,61 @@ export function* resizeSaga(resizeAction: ReduxAction<WidgetResize>) {
...
+    return currentWidgets;
+
+  const reflowingWidgets = reflowState.reflowingWidgets;
+
+  const reflowWidgetKeys = Object.keys(reflowingWidgets || {});
+
+  if (reflowWidgetKeys.length <= 0) return widgets;
+
+  for (const reflowedWidgetId of reflowWidgetKeys) {
+    //eslint-disable-next-line
```
Comment (marks0351) [inline app/client/src/sagas/WidgetOperationSagas.tsx:190]: can we remove the disabler and fix the lint issues.

File: app/client/src/utils/hooks/useCanvasDragging.ts [hunk 9]
```diff
@@ -245,6 +462,92 @@ export const useCanvasDragging = (
...
+              const needsReflow = !(
+                lastSnappedPosition.leftColumn === leftColumn &&
+                lastSnappedPosition.topRow === topRow
+              );
+              lastSnappedPosition = {
+                leftColumn,
+                topRow,
+              };
+              if (canReflow) {
+                if (needsReflow) {
```
Comment (marks0351) [inline app/client/src/utils/hooks/useCanvasDragging.ts:492]: @marks0351 better if this function is moved out to provide better readibility.

File: app/client/src/utils/hooks/useReflow.ts [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+  );
+
+  const prevPositions = useRef<OccupiedSpace | undefined>(
+    originalSpacePosition,
+  );
+  const prevCollidingSpaces = useRef<WidgetCollidingSpaceMap>();
+  const prevMovementMap = useRef<ReflowedSpaceMap>({});
+  // will become a state if we decide that resize should be a "toggle on-demand" feature
+  const shouldResize = true;
+  return function reflowSpaces(
```
Comment (marks0351) [inline app/client/src/utils/hooks/useReflow.ts:51]: @rahulramesha can we provide an interface and use it in useCanvasDragging.

File: app/client/src/utils/reflowHookUtils.ts [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+      if (areIntersecting(widgets[reflowedKey], widgets[widgetId])) {
+        return false;
+      }
+    }
+  }
+
+  return true;
+}
+
+function areIntersecting(r1: FlattenedWidgetProps, r2: FlattenedWidgetProps) {
```
Comment (marks0351) [inline app/client/src/utils/reflowHookUtils.ts:33]: i think this was part of reflowUtils as well. lets get it imported from it.

File: app/client/src/utils/reflowHookUtils.ts [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+import { GridProps, ReflowedSpace, ReflowedSpaceMap } from "reflow/reflowTypes";
+
+export function collisionCheckPostReflow(
+  widgets: {
+    [widgetId: string]: FlattenedWidgetProps;
+  },
+  reflowWidgetKeys: string[],
+  parentId?: string,
+) {
+  const widgetKeys = Object.keys(widgets).filter((widgetId) => {
```
Comment (marks0351) [inline app/client/src/utils/reflowHookUtils.ts:12]: @rahulramesha lets add the boundary check as well.

File: app/client/yarn.lock [hunk 13]
```diff
@@ -14090,7 +14090,8 @@ react-side-effect@^1.1.0:
 
 react-spring@^8.0.27:
   version "8.0.27"
-  resolved "https://registry.npmjs.org/react-spring/-/react-spring-8.0.27.tgz"
+  resolved "https://registry.yarnpkg.com/react-spring/-/react-spring-8.0.27.tgz#97d4dee677f41e0b2adcb696f3839680a3aa356a"
```
Comment (marks0351) [inline app/client/yarn.lock:14093]: @rahulramesha is this needed?

**Review by rahulramesha — APPROVED** [review]

**Pull request merged**